File: python_prototype/clinical_source.py
"""Unified clinical source domain models — mirrors Kotlin ClinicalSource.kt exactly.

Represents clinical citations (articles and tables) with structured section and graphic metadata.
"""

import re
from typing import Optional
from pydantic import BaseModel


class SectionRef(BaseModel):
    section_id: str
    section_title: str


class ClinicalArticle(BaseModel):
    topic_id: str
    topic_title: str
    sections: list[SectionRef] = []

    @property
    def id(self) -> str:
        return self.topic_id

    @property
    def display_title(self) -> str:
        return self.topic_title


class ClinicalTable(BaseModel):
    graphic_id: str
    table_title: str
    parent_topic_id: Optional[str] = None
    parent_topic_title: Optional[str] = None

    @property
    def id(self) -> str:
        return self.graphic_id

    @property
    def display_title(self) -> str:
        return self.table_title


def group_topic_refs_to_articles(refs: list) -> list[ClinicalArticle]:
    """Group flat TopicRef items into structured ClinicalArticle instances with section lists."""
    if not refs:
        return []
    by_topic: dict[str, list] = {}
    for r in refs:
        by_topic.setdefault(r.topic_id, []).append(r)

    articles = []
    for topic_id, group in by_topic.items():
        primary_title = ""
        for item in group:
            if getattr(item, "topic_title", "") and not str(item.topic_title).isdigit():
                primary_title = item.topic_title
                break
        if not primary_title:
            for item in group:
                label = getattr(item, "label", "")
                if label and not str(label).isdigit():
                    primary_title = label
                    break
        if not primary_title:
            primary_title = topic_id

        seen_secs = set()
        sections = []
        for item in group:
            sec_id = getattr(item, "section_id", "")
            if sec_id and sec_id.upper() != "FULL" and sec_id not in seen_secs:
                seen_secs.add(sec_id)
                sections.append(SectionRef(section_id=sec_id, section_title=getattr(item, "label", sec_id)))

        articles.append(
            ClinicalArticle(
                topic_id=topic_id,
                topic_title=primary_title,
                sections=sections,
            )
        )
    return articles


def graphic_refs_to_tables(refs: list, parent_titles: Optional[dict[str, str]] = None) -> list[ClinicalTable]:
    """Convert flat GraphicRef items into ClinicalTable instances."""
    if not refs:
        return []
    parent_titles = parent_titles or {}
    seen = set()
    tables = []
    for r in refs:
        gid = r.graphic_id
        if gid in seen:
            continue
        seen.add(gid)
        p_id = getattr(r, "topic_id", None)
        p_title = parent_titles.get(p_id) if p_id else None
        tables.append(
            ClinicalTable(
                graphic_id=gid,
                table_title=getattr(r, "label", f"Graphic {gid}"),
                parent_topic_id=p_id,
                parent_topic_title=p_title,
            )
        )
    return tables

File: app/src/main/java/com/clinref/app/domain/ai/ClinicalSource.kt
package com.clinref.app.domain.ai

import com.clinref.app.ui.chat.ResolvedGraphicRef
import com.clinref.app.ui.chat.ResolvedTopicRef
import kotlinx.serialization.Serializable

/**
 * Unified domain representation for clinical evidence sources cited in AI responses.
 * Replaces fragmented, ad-hoc reference models with a cohesive, typed hierarchy.
 */
@Serializable
sealed interface ClinicalSource {

    val id: String
    val displayTitle: String

    @Serializable
    data class SectionRef(
        val sectionId: String,
        val sectionTitle: String
    )

    @Serializable
    data class Article(
        val topicId: String,
        val topicTitle: String,
        val sections: List<SectionRef> = emptyList()
    ) : ClinicalSource {
        override val id: String get() = topicId
        override val displayTitle: String get() = topicTitle

        fun toTopicRefs(): List<SafetyValidator.TopicRef> {
            if (sections.isEmpty()) {
                return listOf(
                    SafetyValidator.TopicRef(
                        topicId = topicId,
                        sectionId = "",
                        label = topicTitle,
                        topicTitle = topicTitle
                    )
                )
            }
            return sections.map { sec ->
                SafetyValidator.TopicRef(
                    topicId = topicId,
                    sectionId = sec.sectionId,
                    label = sec.sectionTitle,
                    topicTitle = topicTitle
                )
            }
        }

        fun toResolvedTopicRefs(): List<ResolvedTopicRef> {
            if (sections.isEmpty()) {
                return listOf(
                    ResolvedTopicRef(
                        topicId = topicId,
                        title = topicTitle,
                        sectionId = null,
                        topicTitle = topicTitle
                    )
                )
            }
            return sections.map { sec ->
                ResolvedTopicRef(
                    topicId = topicId,
                    title = sec.sectionTitle,
                    sectionId = sec.sectionId,
                    topicTitle = topicTitle
                )
            }
        }
    }

    @Serializable
    data class Table(
        val graphicId: String,
        val tableTitle: String,
        val parentTopicId: String? = null,
        val parentTopicTitle: String? = null
    ) : ClinicalSource {
        override val id: String get() = graphicId
        override val displayTitle: String get() = tableTitle

        fun toGraphicRef(): SafetyValidator.GraphicRef {
            return SafetyValidator.GraphicRef(
                graphicId = graphicId,
                label = tableTitle,
                topicId = parentTopicId
            )
        }

        fun toResolvedGraphicRef(): ResolvedGraphicRef {
            return ResolvedGraphicRef(
                graphicId = graphicId,
                title = tableTitle,
                topicId = parentTopicId,
                topicTitle = parentTopicTitle
            )
        }
    }

    companion object {
        /**
         * Converts a list of [SafetyValidator.TopicRef] into grouped [ClinicalSource.Article] instances.
         */
        fun fromTopicRefs(refs: List<SafetyValidator.TopicRef>): List<Article> {
            if (refs.isEmpty()) return emptyList()
            return refs.groupBy { it.topicId }.map { (topicId, topicRefsList) ->
                val primaryTitle = topicRefsList.firstOrNull { it.topicTitle.isNotBlank() }?.topicTitle
                    ?: topicRefsList.firstOrNull()?.label?.takeIf { !AiJsonUtils.isNumericOnly(it) }
                    ?: topicId
                val sections = topicRefsList
                    .filter { it.sectionId.isNotBlank() && !it.sectionId.equals("FULL", ignoreCase = true) }
                    .map { SectionRef(sectionId = it.sectionId, sectionTitle = it.label) }
                    .distinctBy { it.sectionId }
                Article(
                    topicId = topicId,
                    topicTitle = primaryTitle,
                    sections = sections
                )
            }
        }

        /**
         * Converts a list of [SafetyValidator.GraphicRef] into [ClinicalSource.Table] instances.
         */
        fun fromGraphicRefs(
            refs: List<SafetyValidator.GraphicRef>,
            parentTopicTitles: Map<String, String> = emptyMap()
        ): List<Table> {
            if (refs.isEmpty()) return emptyList()
            return refs.distinctBy { it.graphicId }.map { ref ->
                val parentTitle = ref.topicId?.let { parentTopicTitles[it] }
                Table(
                    graphicId = ref.graphicId,
                    tableTitle = ref.label,
                    parentTopicId = ref.topicId,
                    parentTopicTitle = parentTitle
                )
            }
        }

        /**
         * Converts UI [ResolvedTopicRef] and [ResolvedGraphicRef] into [ClinicalSource] models.
         */
        fun fromResolved(
            topicRefs: List<ResolvedTopicRef>,
            graphicRefs: List<ResolvedGraphicRef>
        ): Pair<List<Article>, List<Table>> {
            val articles = topicRefs.groupBy { it.topicId }.map { (topicId, group) ->
                val primaryTitle = group.firstOrNull { it.topicTitle.isNotBlank() }?.topicTitle
                    ?: group.firstOrNull()?.title?.takeIf { !AiJsonUtils.isNumericOnly(it) }
                    ?: topicId
                val sections = group
                    .filter { !it.sectionId.isNullOrBlank() && !it.sectionId.equals("FULL", ignoreCase = true) }
                    .map { SectionRef(sectionId = it.sectionId!!, sectionTitle = it.title) }
                    .distinctBy { it.sectionId }
                Article(
                    topicId = topicId,
                    topicTitle = primaryTitle,
                    sections = sections
                )
            }

            val tables = graphicRefs.distinctBy { it.graphicId }.map { ref ->
                Table(
                    graphicId = ref.graphicId,
                    tableTitle = ref.title,
                    parentTopicId = ref.topicId,
                    parentTopicTitle = ref.topicTitle
                )
            }

            return articles to tables
        }
    }
}
